src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,280 @@
+﻿using System;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using NBitcoin.BouncyCastle.Math;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Features.Consensus.Interfaces;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules.ProvenHeaderRules
+{
+    /// <summary>
+    /// Rule to check if the coinstake inside proven header is valid:
+    /// <list type="number">
+    ///     <item><description>Check that coinstake tx has IsCoinStake property equal to true.</description></item>
+    ///     <item><description>Header time is equal to the timestamp of the coinstake tx.</description></item>
+    ///     <item><description>Check if coinstake tx timestamp is divisible by 16 (using timestamp mask).</description></item>
+    ///     <item><description>Verify the coinstake age requirement (*).</description></item>
+    ///     <item>
+    ///     <description>
+    ///         Verify all coinstake transaction inputs (*).
+    ///         <list type="number">
+    ///             <item><description>Input comes from a UTXO.</description></item>
+    ///             <item><description>Verify the ScriptSig</description></item>
+    ///         </list>
+    ///     </description>
+    ///     </item>
+    ///     <item><description>Check if the coinstake kernel hash satisfies the difficulty requirement (*).</description></item>
+    ///     <item><description>Check that coinstake tx is in the merkle tree using the merkle proof.</description></item>
+    ///     <item><description>Verify header signature with the key from coinstake kernel (*).</description></item>
+    /// </list>
+    /// </summary>
+    /// <remarks>(*) - denotes rules that are expensive to execute.</remarks>
+    /// <seealso cref="T:Stratis.Bitcoin.Features.Consensus.Rules.ProvenHeaderRules.ProvenHeaderRuleBase" />
+    public class ProvenHeaderCoinstakeRule : ProvenHeaderRuleBase
+    {
+        /// <summary>The stake validator.</summary>
+        private IStakeValidator stakeValidator;
+
+        /// <summary>PoS block's timestamp mask.</summary>
+        /// <remarks>Used to decrease granularity of timestamp. Supposed to be 2^n-1.</remarks>
+        public const uint StakeTimestampMask = 0x0000000F;
```

Comment:
copypasted value. 

Please make sure there is nothing conensus-related copypasted